[VPD-349] PriceDeviationSentinel Contract

web3rover opened the pull request:
## Summary                                                                                                                                                                                           
                                                                                                                                                                                                       
  - Add **DeviationSentinel** contract that monitors price deviations between ResilientOracle and SentinelOracle, automatically pausing borrow/supply and zeroing collateral factors when large        
  deviations are detected, and restoring them when prices normalize                                                                                                                                    
  - Add **SentinelOracle**, **PancakeSwapOracle**, and **UniswapOracle** as on-chain price sources for deviation comparison                                                                            
  - Add comprehensive unit tests, fork tests, and deployment scripts for BSC testnet                                                                                                                   
                                                                                                                                                                                                       
  ## Problem                                                                                                                                                                                           
                                                                                                                                                                                                       
  Price manipulation or oracle lag can leave markets exposed. The DeviationSentinel provides an automated circuit breaker that a trusted keeper can trigger when oracle prices diverge beyond a        
  configured threshold.                                                                                                                                                                                
                                                                                                                                                                                                       
  ## How It Works                                                                                                                                                                                      
                                                                                                                                                                                                       
  1. **Deviation detected** → Sentinel stores current CF/LT in `marketStates`, sets CF to 0, and pauses supply/borrow                                                                                  
  2. **Deviation resolves** → Sentinel restores stored CF/LT and unpauses actions                                                                                                                      
  3. **Protection against stale state** → If governance creates new e-mode pools or changes CF while the market is paused, the restore logic skips uninitialized pools (`storedCF == 0 && currentCF != 
  0`) and avoids restoring `LT = 0` to prevent liquidation risk                                                                                                                                        
                                                                                                                                                                                                       
  ## DeviationSentinel State Conflicts with Governance Actions                                                                                                                                         
                                                                                                                                                                                                       
  A conflict arises when governance (VIP) needs to modify a market's parameters (CF, LT, pause state, or e-mode configuration) while DeviationSentinel has an active deviation state for that market.  
                                                                                                                                                                                                       
  ### Scenario A — Governance changes CF without calling `resetMarketState`                                                                                                                            
                                                                                                                                                                                                       
  1. Deviation detected → Sentinel stores CF = 80%, sets CF = 0, pauses supply                                                                                                                         
  2. Governance VIP sets CF = 60% (new risk parameter) directly on the comptroller                                                                                                                     
  3. Deviation resolves → Sentinel restores CF = 80% (the stale stored value)                                                                                                                          
  4. **Result:** The governance-intended CF = 60% is silently overwritten with the old 80% value — a security risk where the protocol operates with a higher CF than governance approved               
                                                                                                                                                                                                       
  ### Scenario B — Governance calls `resetMarketState` but doesn't handle pause state                                                                                                                  
                                                                                                                                                                                                       
  1. Deviation detected → Sentinel stores CF = 80%, sets CF = 0, pauses supply                                                                                                                         
  2. Governance calls `resetMarketState` → clears all stored state (`borrowPaused = false`, `cfModifiedAndSupplyPaused = false`, stored CFs cleared)                                                   
  3. Governance VIP sets CF = 60%                                                                                                                                                                      
  4. Deviation resolves → Sentinel sees `borrowPaused = false` and `cfModifiedAndSupplyPaused = false`, so it does nothing                                                                             
  5. **Result:** Supply remains paused with no automated recovery path via the Sentinel if governance didn't manually unpause                                                                          
                                                                                                                                                                                                       
  ### Scenario C — E-mode changes (add/remove pool or market)                                                                                                                                          
                                                                                                                                                                                                       
  1. Sentinel has stored CFs for pool IDs [1, 2, 3]                                                                                                                                                    
  2. Governance creates a new e-mode pool (ID 4) or removes market from pool 2                                                                                                                         
  3. On restore, Sentinel iterates `corePoolId()` to `lastPoolId()` — it may try to restore a CF for a pool the market is no longer in, or skip a newly added pool, leading to incorrect state         
                                                                                                                                                                                                       
  ## Required Governance Procedure                                                                                                                                                                     
                                                                                                                                                                                                       
  When governance needs to intervene on a market with an active deviation state, the VIP **must** follow this exact order:                                                                             
                                                                                                                                                                                                       
                                          
  1. **`resetMarketState(market)`** — Clears all stored state (pause flags, stored CFs/LTs). Note: this clears the sentinel's tracking but does **not** unpause on the comptroller                     
  2. **Make all protocol changes** (new CF/LT values, e-mode pool changes)                                                                                                                             
  3. **Unpause supply/borrow on the comptroller manually** — Since `resetMarketState` only clears sentinel state, governance must explicitly unpause the actions that the sentinel had paused                                                                                                                                      
  4. **`handleDeviation(market)`** — Re-evaluates the market against the current oracle prices with a clean slate                                                                                      
                                                                                                                                                                                                      
  ### Why This Procedure Works in Every Scenario                                                                                                                                                       
                                                                                                                                                                                                       
  By explicitly unpausing and then calling `handleDeviation`, the sentinel captures the fresh governance-intended state:                                                                               
                                                                                                                                                                                                       
  - **Deviation still exists at VIP execution** → `handleDeviation` stores the NEW CF values (set by governance in step 3), sets CF to 0, and re-pauses the market. When the deviation resolves later, 
  the sentinel restores the correct governance-intended CF. ✓                                                                                                                                          
  - **Deviation already resolved by VIP execution** → `handleDeviation` is a no-op since no deviation is detected. The market stays unpaused with the new CF values. ✓                                 
  - **E-mode pool changes (Scenario C)** → `handleDeviation` iterates the updated pool list (`corePoolId()` to `lastPoolId()`) and stores the correct new pool configuration. ✓                        
                                                                                                                                                                                                       
  The market being re-paused when a deviation still exists is the correct behavior — the sentinel should continue protecting the market, just with the updated parameters.                             
                                                                                                                                                                                                       
  ## Contracts Added                                                                                                                                                                                   
                                                                                                                                                                                                       
  | Contract | Description |                                                                                                                                                                           
  |---|---|                                                                                                                                                                                            
  | `DeviationSentinel` | Core sentinel logic — deviation detection, pause/unpause, CF store/restore |                                                                                                 
  | `SentinelOracle` | Aggregates multiple oracle sources for sentinel price comparison |                                                                                                              
  | `PancakeSwapOracle` | TWAP oracle using PancakeSwap V3 pools |                                                                                                                                     
  | `UniswapOracle` | TWAP oracle using Uniswap V3 pools |                                                                                                                                             
                                                                                                                                                                                                       
  ## Test Plan                                                                                                                                                                                         
                                                                                                                                                                                                       
  - [x] Unit tests for DeviationSentinel (deviation handling, CF restore, new pool protection, LT=0 edge case)                                                                                         
  - [x] Unit tests for PancakeSwapOracle, UniswapOracle, SentinelOracle                                                                                                                                
  - [x] Fork tests for DeviationSentinel on BSC                                                                                                                                                        
  - [x] Deployed and verified on BSC testnet
  
  **NOTE:** As discussed with @fred-venus Will work on this [comment](https://github.com/VenusProtocol/venus-periphery/pull/24#discussion_r2706691000) in future versions.

web3rover committed:
feat: added core logic

web3rover committed:
fix: usd price should be 18 decimals

web3rover committed:
fix: handle emode groups

web3rover committed:
fix: use pancakeswap v3

web3rover committed:
fix: updated package

web3rover committed:
fix: fixed pancakeswap calculation

web3rover committed:
fix: optimised code

web3rover committed:
fix: added uniswap test

web3rover committed:
fix: added fork tests for deviation logic

fred-venus commented on contracts/PriceDeviationSentinel/PriceDeviationSentinel.sol:
```diff
@@ -0,0 +1,499 @@
+        }
+
+        deviationPercent = (priceDiff * 100) / oraclePrice;
+        hasDeviation = deviationPercent > config.deviation;
```
maybe ">="  is better ?

fred-venus commented on contracts/PriceDeviationSentinel/PriceDeviationSentinel.sol:
```diff
@@ -0,0 +1,499 @@
+                _pauseBorrow(market, IComptroller.Action.BORROW);
+                state.isPaused = true;
+            } else {
+                _setCollateralFactorToZero(market);
```
Probably pause supply as well ? if supply is not paused, i was thinking ppl could still leverage this and supplying more to protect their existing poistion being liquidated as the liquidation threshold is not changed

web3rover committed:
fix: enable or disable markets

web3rover committed:
fix: change deviation logic

web3rover commented on contracts/PriceDeviationSentinel/PriceDeviationSentinel.sol:
```diff
@@ -0,0 +1,499 @@
+    /// @notice Get USD price of a token from DEX using stored configuration
+    /// @param token Token address to get price for
+    /// @return price USD price in 18 decimals
+    function getDexPrice(address token) public view returns (uint256 price) {
```
So basically we will create a contract SentinelOracle, which internally fetches prices from oracles like PancakeSwapOracle, UniswapOracle and so on contracts based on configuration.

So PriceDeviationSentinel will query SentinelOracle and ReslientOracle both for the price of an token and check for deviation.

wdyt?

web3rover replied:
Fixed

web3rover committed:
fix: pause supply

web3rover committed:
fix: early return

web3rover commented on contracts/PriceDeviationSentinel/PriceDeviationSentinel.sol:
```diff
@@ -0,0 +1,499 @@
+
+        MarketState storage state = marketStates[address(market)];
+
+        if (hasDeviation) {
```
Yes fixed

web3rover commented on contracts/PriceDeviationSentinel/PriceDeviationSentinel.sol:
```diff
@@ -0,0 +1,499 @@
+        address underlyingToken = market.underlying();
+        DeviationConfig memory config = tokenConfigs[underlyingToken];
+
+        uint256 oraclePriceRaw = RESILIENT_ORACLE.getPrice(underlyingToken);
```
If the resilient oracle reverts then the whole transaction will revert so it won't pause

web3rover committed:
fix: create seperate oracle contracts

web3rover committed:
fix: reanmed contract

web3rover committed:
fix: move oracles to oracle repo

web3rover committed:
Revert "fix: move oracles to oracle repo"

This reverts commit cffb86f8ef93832e38c823e38ee4a268f472fab4.

web3rover committed:
fix: fixed e2e tests

web3rover committed:
fix: fixed tests

web3rover committed:
feat: added unit tests

Debugger022 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,439 @@
+    }
+
+    /// @notice Handle price deviation for a market by pausing or adjusting collateral factor
+    /// @param market The vToken market to handle
```
```suggestion
    /// @param market The vToken market to handle
    /// @custom:error UnauthorizedKeeper Thrown when caller is not an authorized keeper
```

Debugger022 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,439 @@
+        } else {
+            // Isolated pool
+            uint256 originalLT = state.originalLT;
+            IILComptroller(address(comptroller)).setCollateralFactor(address(market), originalCF, originalLT);
```
We can directly use `state.originalCF` and `state.originalLT`. It will save space.

web3rover committed:
chore: added events and custom errors

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,439 @@
+
+    /// @notice Handle price deviation for a market by pausing or adjusting collateral factor
+    /// @param market The vToken market to handle
+    function handleDeviation(IVToken market) external onlyKeeper {
```
Fixed

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,439 @@
+
+        if (token == address(0)) revert ZeroAddress();
+        if (config.deviation == 0) revert ZeroDeviation();
+        if (config.deviation > MAX_DEVIATION) revert ExceedsMaxDeviation();
```
Yes. After setting deviation we can optionally enable or disable the market for monitoring.

web3rover committed:
fix: fix LT

web3rover committed:
fix: fixed var name

web3rover committed:
fix: optimise code

web3rover committed:
fix: optimise code

web3rover committed:
fix: added deployments

web3rover committed:
fix: merge conflict

web3rover committed:
fix: added deployments

web3rover committed:
feat: updating deployment files

Debugger022 commented on contracts/Oracles/PancakeSwapOracle.sol:
```diff
@@ -0,0 +1,141 @@
+    /// @return price Price in (36 - token decimals) format
+    /// @custom:error InvalidPool is thrown when pool address is zero or token not in pool
+    function _getPancakeSwapV3Price(address pool, address token) internal view returns (uint256 price) {
+        if (pool == address(0)) revert InvalidPool();
```
This check is not necessary here, as we are already performing it in the external function.

Debugger022 replied:
Same comments for `UniswapOracle`.

fred-venus commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,469 @@
+            if (state.supplyPaused) {
+                _unpauseSupply(market, IComptroller.Action.MINT);
+                state.supplyPaused = false;
+            }
```
if it's either of the following options
- state.cfModified but !state.supplyPaused
- !state.cfModified but state.supplyPaused

I would say there must be some setting submitted by some other source, in that case, i prefer just siting doing nothing and wait manually taking over

fred-venus replied:
After a second look, actually should we just combine cfModified and supplyPaused as i dont think there would be some case that those 2 variables having different value

web3rover replied:
Combined both variables

fred-venus commented on contracts/Oracles/PancakeSwapOracle.sol:
let's put oracles under DeviationSentinel, it's unlikely to be used by other features

web3rover committed:
fix: added imports

web3rover committed:
Merge branch 'feat/vpd-349' of github.com:VenusProtocol/venus-periphery into feat/vpd-349

web3rover committed:
Merge branch 'develop' into feat/vpd-349

web3rover committed:
fix: remove duplicate imports

web3rover committed:
fix: optimise code

web3rover committed:
fix: remove unwanted error

web3rover committed:
fix: optimise code

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,469 @@
+                if (state.cfModified && state.supplyPaused) return;
+
+                _setCollateralFactorToZero(market);
+                _pauseSupply(market, IComptroller.Action.MINT);
```
Fixed

web3rover committed:
fix: fixed interfaces

web3rover committed:
fix: rearrange files

web3rover committed:
fix: use less vars

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,469 @@
+            IILComptroller.Market memory marketData = IILComptroller(address(comptroller)).markets(address(market));
+            if (marketData.isListed) {
+                state.originalCF = marketData.collateralFactorMantissa;
+                state.originalLT = marketData.liquidationThresholdMantissa;
```
Yes it will reduce the number of variables. Fixed.

web3rover committed:
fix: fixed tests

web3rover committed:
fix: added reset func

web3rover committed:
fix: fixed lint

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,469 @@
+    /// @param originalLT Original liquidation threshold before modification (for IL pools)
+    /// @param poolCFs Mapping of pool ID to original collateral factor (for core pool)
+    /// @param poolLTs Mapping of pool ID to original liquidation threshold (for core pool)
+    struct MarketState {
```
Added a reset function

web3rover committed:
fix: fixed decimals

fred-venus commented on contracts/DeviationSentinel/Oracles/PancakeSwapOracle.sol:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        uint256 referencePrice = RESILIENT_ORACLE.getPrice(referenceToken);
+        uint8 targetDecimals = IERC20Metadata(targetToken).decimals();
```
unused `targetDecimals` and `referenceDecimals`

web3rover replied:
Fixed

web3rover committed:
fix: optimise code

web3rover commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,490 @@
+    function _setCollateralFactorToZero(IVToken market) internal {
+        IComptroller comptroller = market.comptroller();
+        MarketState storage state = marketStates[address(market)];
+
```
Fixed

web3rover committed:
fix: optimise code

web3rover committed:
fix: removed unused vars

web3rover committed:
fix: added test case for non 18 reference token decimals

web3rover commented on contracts/DeviationSentinel/Oracles/UniswapOracle.sol:
```diff
@@ -118,20 +117,6 @@ contract UniswapOracle is AccessControlledV8 {
-                price = price / (10 ** (18 - targetPriceDecimals));
-            }
-        }
+        price = FullMath.mulDiv(referencePrice, (10 ** 18), targetTokensPerReferenceToken);
```
Fixed

web3rover committed:
fix: deployed new contracts

web3rover committed:
fix: vld-02

web3rover committed:
fix: vld-03

web3rover committed:
fix: vld-08

web3rover committed:
fix: vld-09

web3rover committed:
fix: vld-12

web3rover committed:
fix: vld-06

web3rover committed:
fix: removed comments

web3rover committed:
fix: removed comments

web3rover committed:
fix: vld-03

web3rover committed:
fix: added comments to call resetMarketState

web3rover committed:
fix: m01

web3rover committed:
fix: i02

web3rover committed:
fix: i04

Debugger022 committed:
feat: add comment

Debugger022 commented on contracts/DeviationSentinel/Oracles/UniswapOracle.sol:
```diff
@@ -0,0 +1,120 @@
+
+        uint256 priceX96 = FullMath.mulDiv(sqrtPriceX96, sqrtPriceX96, FixedPoint96.Q96);
+
+        address targetToken = token;
```
unused variable

fred-venus replied:
agree, let's remove it

Debugger022 replied:
[fixed](https://github.com/VenusProtocol/venus-periphery/pull/44/changes/b209db75bf497783af22d4d973c39d3becbae313#diff-f6eb4e09f9e67c6fe1a099730a1695fc52145dcac8f97403e0e074c075c23fd3L1)

Debugger022 commented on contracts/DeviationSentinel/Oracles/PancakeSwapOracle.sol:
```diff
@@ -0,0 +1,120 @@
+
+        uint256 priceX96 = FullMath.mulDiv(sqrtPriceX96, sqrtPriceX96, FixedPoint96.Q96);
+
+        address targetToken = token;
```
unused variable

fred-venus replied:
let's remove it as well

Debugger022 replied:
[fixed](https://github.com/VenusProtocol/venus-periphery/pull/44/changes/b209db75bf497783af22d4d973c39d3becbae313#diff-f6eb4e09f9e67c6fe1a099730a1695fc52145dcac8f97403e0e074c075c23fd3L1)

Debugger022 commented on contracts/DeviationSentinel/Oracles/PancakeSwapOracle.sol:
```diff
@@ -0,0 +1,120 @@
+    ResilientOracleInterface public immutable RESILIENT_ORACLE;
+
+    /// @notice Mapping of token addresses to their pool addresses
+    mapping(address => address) public tokenPools;
```
instead of this we could use `getPool`  method of  pancakeSwap

fred-venus replied:
let's use our own mapping here as it will require more params like fee tier etc to decide the pool addr through pcs method

Debugger022 commented on contracts/DeviationSentinel/Oracles/UniswapOracle.sol:
```diff
@@ -0,0 +1,120 @@
+    ResilientOracleInterface public immutable RESILIENT_ORACLE;
+
+    /// @notice Mapping of token addresses to their pool addresses
+    mapping(address => address) public tokenPools;
```
instead of this we could use `getPool`  method of  uniSwap

fred-venus replied:
same as above

Debugger022 commented on contracts/DeviationSentinel/Oracles/SentinelOracle.sol:
```diff
@@ -0,0 +1,71 @@
+    /// @param asset Address of the asset
+    /// @return price Price in (36 - asset decimals) format, same as ResilientOracle
+    /// @custom:error TokenNotConfigured is thrown when asset has no oracle configured
+    function getPrice(address asset) external view returns (uint256 price) {
```
Both the PancakeSwapOracle and UniswapOracle contracts share the same core logic, so we could consolidate them into a single common oracle contract instead of maintaining two separate ones.

Additionally, we could further improve security by aggregating and comparing price data from multiple DEXs, rather than relying on a single source.

fred-venus replied:
Yeah, i agree both points. That's for sure something we could do,  together with what u mentioned TokenConfig has only one field, i am thinking we can have more fields:

we could support amm pools array for one asset and each pool attached with a weight, and get price can simply return:
`price1 * weight1 + price2 * weight2` etc..

wdty ?

Debugger022 replied:
yeah it's sounds good

Debugger022 commented on contracts/DeviationSentinel/Oracles/SentinelOracle.sol:
```diff
@@ -0,0 +1,71 @@
+contract SentinelOracle is AccessControlledV8 {
+    /// @notice Configuration for token price source
+    /// @param oracle Address of the DEX oracle to use for this token
+    struct TokenConfig {
```
why we are having struct for one field

Debugger022 replied:
@fred-venus

fred-venus replied:
Like mentioned here https://github.com/VenusProtocol/venus-periphery/pull/24#discussion_r2706691000, original thinking is to have several price source

Debugger022 committed:
chore: cleanup unused code and add SPDX identifier

Debugger022 committed:
fix: unify CollateralFactorUpdated event and expand test coverage

- Remove duplicate CollateralFactorUpdated event overload for isolated pools,
  consolidate into single event with poolId param
- Rewrite DeviationSentinel unit tests with comprehensive coverage: zero-address
  constructor checks, emode pool iteration, CF caching/restore, deviation
  direction flips, and sentinel direct price scenarios
- Expand PancakeSwapOracle and UniswapOracle test suites with additional edge cases
- Add thorough fork test suite covering end-to-end deviation handling on BSC mainnet

Debugger022 committed:
Merge pull request #44 from VenusProtocol/feat/vpd-442

[VPD-442] Hashdit Findings for Emergency Pause

Debugger022 committed:
Merge pull request #42 from VenusProtocol/feat/vpd-432

[VPD-432] Certik Audit Fixes for Emergency Pause

Debugger022 committed:
Merge branch 'feat/vpd-402' into feat/vpd-349

Debugger022 committed:
chore: fixed Deviation testing

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
chore: fixed linter error

Debugger022 committed:
chore: retrigger ci

Debugger022 committed:
fix: skip waitConfirmations on local hardhat network

Debugger022 committed:
feat: add bscmainnet deployments

Debugger022 committed:
fix: use deployer as proxy owner on non-live networks in sentinel deploy

Conditionally set proxy owner to deployer instead of timelock on local
hardhat networks for PancakeSwapOracle and UniswapOracle, matching the
pattern already used by SentinelOracle and DeviationSentinel. Also
remove waitConfirmations and inline verify calls from the deploy script.

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
feat: refactor DeviationSentinel fork tests to use mainnet contracts

- Replace mock-based tests with actual BSC mainnet fork tests
- Use deployed contract addresses from bscmainnet_addresses.json
- Execute VIP-900 to set up proper permissions during test setup
- Extend Chainlink staleness period for BTCB to handle fork test timing
- Cache oracle price at setup to avoid staleness issues during tests
- Change test asset from vBNB to vBTCB (vBNB wraps native BNB with no ERC20 underlying)
- Add mocha timeout configuration (200s) for fork tests in hardhat.config.ts
- Fix linter errors for unused variables

GitGuru7 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,518 @@
+    /// @param isTrusted Whether the keeper should be trusted
+    /// @custom:event Emits TrustedKeeperUpdated event
+    /// @custom:error ZeroAddress is thrown when keeper address is zero
+    function setTrustedKeeper(address keeper, bool isTrusted) external {
```
we can consider reverting when isTrusted is unchanged to avoid unnecessary storage writes and also emit the previous value for better tracking.

GitGuru7 replied:
applicable to other similar setters

Debugger022 replied:
The current implementation is correct and functional. Your suggestions are valid optimizations:                                                                                                                                  
  1. Reverting on unchanged value saves gas by avoiding redundant `SSTORE` operations                                                                                                                                                
  2. Emitting previous value improves off-chain tracking and event indexing                                                                                                                                                        
                                                                                                                                                                                                                                   
  However, the current simpler approach has benefits:                                                                                                                                                                              
  - Idempotent behavior (no unexpected reverts)                                                                                                                                                                                    
  - Less code complexity                                                                                                                                                                                                           
  - Callers don't need to check current state before calling                                                                                                                                                                       
                                                                                                                                                                                                                                   
Since the setter will primarily be called by a whitelisted address, we can assume incorrect parameters won’t be sent. Given that the code is already deployed on mainnet, I’d suggest skipping this. @fred-venus WDYT?T?

fred-venus replied:
i am open, can skip

GitGuru7 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,518 @@
+                    //   so we skip restoring to avoid overwriting new pool config with zero values.
+                    // - If storedLT is 0, skip restoration to prevent setting LT=0, which could cause immediate liquidation risk.
+                    //   This also protects against uninitialized storage for new pools.
+                    if (storedCF == 0 && currentCF != 0) {
```
Can we consider avoiding the `storedCF == 0 `check and rely only on currentCF != 0 when restoring values? If a CF is updated via a VIP during the pause, should the sentinel still reset it, or is the current behavior intentional ?

Debugger022 replied:
The check is correct and intentional. It distinguishes between:                                                                                                                                                                  
  1. `storedCF == 0` because never stored (new pool created during pause) → skip if `currentCF != 0` to preserve governance's new pool config                                                                                          
  2. storedCF == 0 because original CF was 0 → restore allowed if currentCF is also 0                                                                                                                                              
                                                                                                                                                                                                                                   
 See line [475-477](https://github.com/VenusProtocol/venus-periphery/blob/f60beefd73345c6af40d14028a5c6fdbb6bbbe5f/contracts/DeviationSentinel/DeviationSentinel.sol#L475-L477) comments: 

**If storedCF is 0 and currentCF != 0, this pool was added after _setCollateralFactorToZero, so we skip restoring to avoid overwriting new pool config with zero values.**

GitGuru7 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,518 @@
+    event SupplyUnpaused(address indexed market);
+
+    /// @notice Emitted when collateral factor is updated
+    /// @notice Emitted when collateral factor is updated
```
```suggestion
```

Debugger022 replied:
As the code is already deployed on mainnet, I would suggest to skip this @fred-venus WDYT?

fred-venus replied:
yeah, can skip

GitGuru7 commented on contracts/DeviationSentinel/DeviationSentinel.sol:
```diff
@@ -0,0 +1,518 @@
+                    uint256 result = CORE_POOL_COMPTROLLER.setCollateralFactor(i, address(market), storedCF, storedLT);
+                    if (result != 0) revert ComptrollerError(result);
+
+                    emit CollateralFactorUpdated(address(market), i, 0, storedCF);
```
On restore, the event assumes currentCF is always 0, but if governance updates the CF during the pause, this may not hold. Should we consider emitting the actual previous value instead (from currentCF) for better accuracy and traceability?

Debugger022 replied:
The current behavior is intentional. The event tracks DeviationSentinel's state transitions:                                                                                                                                     
  - On pause: oldCF = original CF → newCF = 0 (sentinel zeroing)                                                                                                                                                                   
  - On restore: oldCF = 0 → newCF = storedCF (sentinel restoring from what it set)                                                                                                                                                 
                                                                                                                                                                                                                                   
If governance modifies CF during the pause, that's a separate action outside the sentinel's scope. The sentinel's event accurately reflects: "I'm restoring from 0 (what I set) to the stored value.
  
Does this make sense?

GitGuru7 commented on contracts/DeviationSentinel/Oracles/PancakeSwapOracle.sol:
```diff
@@ -0,0 +1,111 @@
+ * @author Venus
+ * @notice Oracle contract for fetching asset prices from PancakeSwap V3
+ */
+contract PancakeSwapOracle is AccessControlledV8 {
```
Nit: We could consider using a common abstract/base contract here to avoid duplicating the same logic.

Debugger022 replied:
As discussed with @fred-venus in https://github.com/VenusProtocol/venus-periphery/pull/24#discussion_r2706691000, we plan to enhance the oracle architecture in future versions to support multiple AMM pools  
  with weighted price aggregation. I'll address the base contract consolidation as part of that work.

GitGuru7 reviewed (commented):
implementation LGTM, left few nitpicks

Debugger022 committed:
feat: add emergency pause hashdit final audit report

fred-venus commented:
merge as vip has will be proposed soon

fred-venus commented:
cc @Debugger022 , once the audit report is available, feel free to create a new pr
